torch/utils/data/distributed.py

```diff
@@ -61,7 +73,14 @@ def __iter__(self):
         return iter(indices)
 
     def __len__(self):
-        return self.num_samples
+        if self.padding:
+            return self.num_samples
+        return self._dataset_length
```

Comment:
This seems strange. Why isn't this also num samples?

Reply:
Indeed. Should `_dataset_length` just be replaced by `num_samples`?

Reply:
`DistributedSample`._len__ will typically be used by map style datasets only. given an index list such as `[1,2,3,4,5,6,7]` and `num_replicas=3` as input, sampler should return `[[1,2,3],[4,5,6],[7,1,2]]` and `len(sampler)=3` for every batch

When `padding=False`, on the other hand, given the same output, the sampler must return `[1,2,3],[4,5,6],[7]` but its length varies and for `IterableDataset` not really used.

Returning `self.num_samples` works for me. I can add a note informing that regardless of `padding` value, :meth:`__len__` will return `ceil(dataset length / num_replicas)`

Reply:
Sounds good to me! - Did a missclick on the frowny face....